### PR TITLE
Change AMQP API to client control retries

### LINKED ISSUE #27: Invalid events are remaining in the message queue
The invalid messages should be removed from the message queue since it won't be used. This is causing the log to get dirty and the message queue to use memory unnecessarily.

```
(node:31382) UnhandledPromiseRejectionWarning: Error: Unknown event type cloud.device.auth
    at MessageHandler.getHandler (/Users/jneto/workspace/knot/gw/knot-fog-connector/src/network/MessageHandler.js:51:13)
    at MessageHandler.getHandler [as handleMessage] (/Users/jneto/workspace/knot/gw/knot-fog-connector/src/network/MessageHandler.js:70:26)
    at Channel.BaseChannel.dispatchMessage (/Users/jneto/workspace/knot/gw/knot-fog-connector/node_modules/amqplib/lib/channel.js:484:12)
    at Channel.BaseChannel.handleDelivery (/Users/jneto/workspace/knot/gw/knot-fog-connector/node_modules/amqplib/lib/channel.js:493:15)
    at Channel.emit (events.js:198:13)
    at /Users/jneto/workspace/knot/gw/knot-fog-connector/node_modules/amqplib/lib/channel.js:279:10
    at Channel.content (/Users/jneto/workspace/knot/gw/knot-fog-connector/node_modules/amqplib/lib/channel.js:332:9)
    at Channel.C.acceptMessageFrame (/Users/jneto/workspace/knot/gw/knot-fog-connector/node_modules/amqplib/lib/channel.js:247:31)
    at Channel.C.accept (/Users/jneto/workspace/knot/gw/knot-fog-connector/node_modules/amqplib/lib/channel.js:400:17)
    at Connection.mainAccept [as accept] (/Users/jneto/workspace/knot/gw/knot-fog-connector/node_modules/amqplib/lib/connection.js:64:33)
    at Socket.go (/Users/jneto/workspace/knot/gw/knot-fog-connector/node_modules/amqplib/lib/connection.js:478:48)
    at Socket.emit (events.js:198:13)
    at emitReadable_ (_stream_readable.js:539:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
(node:31382) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:31382) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

### LINKED ISSUE #35: Publish data is always kept on queue when device is unregistered
If the thing is unregistered and the thing keeps sending the data, the queue should reject the messages and not just nack them.

### LINKED ISSUE #36: The message device.register is being requeued
The message `device.register` is not rejecting when an error occurs. The message `device.registered` should reply an error-reason or error-code to the client.

### PR DESCRIPTION
**Pull request by ramonhpr**
What does this implement/fix? Explain your changes.
---------------------------------------------------
This patchset changes the API to avoid that a message is re-queued when there is some error on connector library. Now the client that sent the message receives the message with the error to handle it.

Does this close any currently open issues?
------------------------------------------
Fixes #36, Fixes #27, Fixes #35 

Any relevant logs, error output, etc?
-------------------------------------
No

Where has this been tested?
---------------------------

**Operating System/Platform:** 4.15.0-70-generic #79-Ubuntu SMP Tue Nov 12 10:36:11 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

**NPM Version:** 6.7.0

**NodeJS Version:** 11.13

### CODE REVIEW

File: docs/api/amqp.md [hunk 1]
```diff
@@ -99,24 +99,36 @@ Message with the list of devices registered on the cloud.
...
+    {
+      "id":"3aa21010cda96fe9",
+      "name":"KNoT Dongle",
+      "schema":[
+        {
+          "sensor_id":0,
+          "value_type":3,
+          "unit":0,
+          "type_id":65521,
+          "name":"LED"
```
Comment (humbertowanderley) [inline docs/api/amqp.md]: Please, add white space between key and value.

File: docs/api/amqp.md [hunk 2]
```diff
@@ -165,12 +185,23 @@ JSON in the following format:
     * `id` **String** device ID
     * `token` **String** device token
 
-#### Example
+#### Success Response Example
 
 ```json
 {
   "id":"3aa21010cda96fe9",
   "token":"5b67ce6bef21701331152d6297e1bd2b22f91787",
```
Comment (humbertowanderley) [inline docs/api/amqp.md]: Same here.

File: src/interactors/ListDevices.js [hunk 3]
```diff
@@ -7,12 +7,19 @@ class ListDevices {
...
+      await this.queue.sendList({
+        devices: devices.map((dev) => {
+          const device = dev;
+          device.schema = convertToSnakeCase(dev.schema);
+          return device;
+        }),
+        error: null,
+      });
+    } catch (error) {
+      this.queue.sendList({ devices: [], error: error.message });
```
Comment (humbertowanderley) [inline src/interactors/ListDevices.js]: Please, insert `await`  before the `sendList( )` call.

File: docs/api/amqp.md (hunk 1 above)
Comment (ramonhpr) [reply docs/api/amqp.md]: Done

File: src/interactors/ListDevices.js (hunk 3 above)
Comment (ramonhpr) [reply src/interactors/ListDevices.js]: Done

File: docs/api/amqp.md (hunk 2 above)
Comment (ramonhpr) [reply docs/api/amqp.md]: Done

File: src/network/MessagePublisher.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
+const exchangeName = 'fog';
+const exchangeControl = 'control';
+
+export default class MessagePublisher {
```
Comment (netoax) [inline src/network/MessagePublisher.js]: I'd prefer to move the `export` expression to the end of the file in order to follow the pattern of the other classes.

File: src/network/MessagePublisher.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  constructor(queue) {
+    this.queue = queue;
+  }
+
+  async sendRegister(body) {
```
Comment (netoax) [inline src/network/MessagePublisher.js]: This method is sending an event to the queue: 'device.registered', maybe its name should be `sendRegisteredDevice`. I think `sendRegister` can lead to confusion when distinguishing between a command and an event.

Comment (netoax) [reply src/network/MessagePublisher.js]: This observation is also applied to the other methods.

File: src/interactors/UnregisterDevice.js [hunk 6]
```diff
@@ -1,15 +1,21 @@
 import logger from 'util/logger';
 
 class UnregisterDevice {
-  constructor(deviceStore, cloudConnector) {
+  constructor(deviceStore, cloudConnector, queue) {
```
Comment (netoax) [inline src/interactors/UnregisterDevice.js]: What do you think about receiving the parameter `queue` as `publisher` instead?

File: docs/api/amqp.md [hunk 7]
```diff
@@ -147,17 +147,25 @@ Message with the status of device authentication command.
...
+  "error": null
 }
 ```
 
+#### Error Response Example
+
+```json
+{
+  "id": "3aa21010cda96fe9",
+  "error": "Not authorized"
```
Comment (netoax) [inline docs/api/amqp.md]: Unauthorized

**Review by netoax — CHANGES_REQUESTED** [review]

**Comment by netoax** [comment]
I didn't understand the commit _4a8ee46bc615fc35ec47050bf957c62e6d8ec5e6_ based on its commit title. Please, change it to something like "Change northbound exchange name", "Change cloud message exchange name". Details can go in the commit message.

**Comment by netoax** [comment]
Couldn't the commit _dcb9934e0ba92cc3b93e934225028a8874071fe3_ title be like the _a81a1bd4519e2863ff9efdf6b9196e75b4685f46_ and _0eb02a4542b6df2b62991abb852d864914d694f3_ ones?

Also, `error` can be written without a single quote and maybe as the commit messages are equals they can be unnecessary.

**Review by humbertowanderley — APPROVED** [review]

File: src/network/MessagePublisher.js (hunk 4 above)
Comment (ramonhpr) [reply src/network/MessagePublisher.js]: Done

File: src/network/MessagePublisher.js (hunk 5 above)
Comment (ramonhpr) [reply src/network/MessagePublisher.js]: Done

File: docs/api/amqp.md (hunk 7 above)
Comment (ramonhpr) [reply docs/api/amqp.md]: Done

**Comment by netoax** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/network/MessagePublisher.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/joaoaneto/knot-fog-connector/pullRequest?prid=4709148&bid=15729415)

File: src/interactors/UnregisterDevice.js (hunk 6 above)
Comment (ramonhpr) [reply src/interactors/UnregisterDevice.js]: Done

Comment (netoax) [reply src/interactors/UnregisterDevice.js]: Can you do this change for the other interactors too?

File: docs/api/amqp.md [hunk 8]
```diff
@@ -25,7 +25,8 @@ This document specifies the messages expected by the connector through a publish
 
 ### Update data
 
-Updates a thing's sensor value.
+Updates a thing's sensor value. This message will be on the queue for 10 seconds
+befor expires.
```
Comment (netoax) [inline docs/api/amqp.md]: **before**

File: docs/api/amqp.md [hunk 9]
```diff
@@ -57,7 +58,8 @@ JSON in the following format:
 
 ### Request data
 
-Request a thing's sensor value.
+Request a thing's sensor value. This message will be on the queue for 10 seconds
+befor expires.
```
Comment (netoax) [inline docs/api/amqp.md]: same here

**Review by netoax — CHANGES_REQUESTED** [review]

File: src/interactors/UnregisterDevice.js (hunk 6 above)
Comment (ramonhpr) [reply src/interactors/UnregisterDevice.js]: Done

File: docs/api/amqp.md (hunk 8 above)
Comment (ramonhpr) [reply docs/api/amqp.md]: done

File: docs/api/amqp.md (hunk 9 above)
Comment (ramonhpr) [reply docs/api/amqp.md]: done

**Review by netoax — APPROVED** [review]

**Pull request merged**